server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,581 @@
+                                    rhier.add_unexpected_entries(
+                                        Hierarchy.BAD_PREFIXES, controller, path
+                                    )
+                            elif not prefix_file.exists():
+                                rhier.add_unexpected_entries(
+                                    Hierarchy.MISSING_PREFIX_FILES, controller, path
+                                )
+                            else:
+                                f = 0
+                                try:
+                                    with prefix_file.open(mode="r") as file:
+                                        prefix = file.read().replace("\n", "")
+                                except Exception:
+                                    f = 1
+                                if f == 1:
+                                    rhier.add_unexpected_entries(
+                                        Hierarchy.BAD_PREFIX_FILES, controller, path
+                                    )
+                                else:
+                                    if prefix != prefix_path:
+                                        rhier.add_unexpected_entries(
+                                            Hierarchy.BAD_PREFIXES, controller, path
+                                        )
+                        verify_user_arg(rhier, mdlogcfg, controller, path)
+    return 0
+
+
+# Controller Hierarchy
+def verify_controllers(hier):
+    """ 'hier' is a ControllerHierarchy class object. """
+    controllers = Path(hier.path).glob("*")
+    verify_valid_controllers(hier, controllers)
+
+    for controller in hier.controllers:
+        dir_path = Path(hier.path, controller)
+        unexpected_dirs = list()
+        if not Path(hier.config.ARCHIVE, controller).is_dir():
+            """We have a controller in the hierarchy which does not have a
+            controller of the same name in the archive hierarchy.  All
+            we do is report it, don't bother analyzing it further.
+            """
+            hier.add_controller_list(Hierarchy.MIALIST, Path(controller).name)
+        else:
+            """Report any controllers with objects other than directories
+            and links, while also recording any empty controllers.
+            """
+            count_dir_entries = 0
+            direct_entries = dir_path.glob("*")
+            for item_path in direct_entries:
+                count_dir_entries += 1
+                if not item_path.is_dir() and not item_path.is_symlink():
+                    unexpected_dirs.append(controller)
+                    break
+            else:
+                if count_dir_entries == 0:
+                    hier.add_controller_list(Hierarchy.EMPTY_CONTROLLERS, controller)
+                    continue
+
+            if unexpected_dirs:
+                hier.add_controller_list(Hierarchy.UNEXPECTED_CONTROLLERS, controller)
+            hier.add_verify_list(controller)
```

Comment:
Alternatively:
```
            # Report any controllers with objects other than directories
            # and links, while also recording any empty controllers.
            direct_entries = dir_path.glob("*")
            if len(direct_entries) == 0:
                hier.add_controller_list(Hierarchy.EMPTY_CONTROLLERS, controller)
            else:
                for item_path in direct_entries:
                    if not item_path.is_dir() and not item_path.is_symlink():
                        hier.add_controller_list(Hierarchy.UNEXPECTED_CONTROLLERS, controller)
                        break                    
                hier.add_verify_list(controller)
```
With this change, we don't need the `unexpected_dirs` list, which we seem to be using as a boolean....

Reply:
Thank you :). This code seems to be changing some test cases. I will go through it again

Reply:
I could easily have missed a case.  🙈

Reply:
oh, sorry it was my bad I didn't realize we cannot go through the elements of a generator twice.

Reply:
Yes:  generators are handy in many ways, the foremost being that you don't have to instantiate the entire list up front and store it in memory; however, they do have the downside that you cannot move them backwards, you cannot access them randomly (via an index), and if you want to "rewind" them you have to recreate them.